### ISSUE TITLE
[naming-convention] Multiple objects with the same selector, but different modifiers do not work.

### ISSUE THREAD

**Issue by infacto** [typescript-eslint/typescript-eslint#2711]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I really like this clean code rule. Sorry if this is my fault. However, it appears that several objects are not respected in the [naming-convention](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md). 

**Repro**

[**Download minimal project**](https://github.com/typescript-eslint/typescript-eslint/files/5439620/project.zip) (Really small demo to reproduce.)

```JSON
{
  "rules": {
"@typescript-eslint/naming-convention": [
      "warn",
      {
        "selector": "variable",
        "modifiers": ["const"],
        "format": ["strictCamelCase"]
      },
      {
        "selector": "property",
        "format": ["strictCamelCase"],
        "leadingUnderscore": "allow",
        "filter": {
          "regex": "[- ]",
          "match": false
        }
      },
      {
        "selector": "property",
        "modifiers": ["readonly"],
        "format": ["strictCamelCase", "UPPER_CASE"]
      },
      {
        "selector": "parameter",
        "format": ["strictCamelCase"],
        "leadingUnderscore": "allow"
      },
      {
        "selector": "typeLike",
        "format": ["StrictPascalCase"]
      },
      {
        "selector": "enumMember",
        "format": ["StrictPascalCase"]
      }
    ]
  }
}
```

```TS
export class MyTest {
  private static readonly SOME_VALUE = 42;

  constructor() {
    console.log(MyTest.SOME_VALUE);
  }
}
```

**Expected Result**

It should support multiple objects with same selector but different modifiers to set format rules on each case.

**Actual Result**

The one "property" selector entry is dominant. I cannot use more than one object with this selector but different modifiers.
Is this not supported? Then it should be (feature request). Or either this is a bug, or I doing something wrong. But what?

**Additional Info**

![demo](https://user-images.githubusercontent.com/60390085/97191333-112bea00-17a7-11eb-9997-a402b585a0a0.gif)

I changed the order in array. But does not help. Is the order relevant? Should it? I think yes. Maybe this is the problem? 🤔 An object with modifiers should have higher prio than others. Just thinking aloud... But the order is important. Anyway, any guesses what the problem is?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `10.22.1` |

VS Code: 

```text
Version: 1.50.1 (user setup)
Commit: d2e414d9e4239a252d1ab117bd7067f125afd80a
Date: 2020-10-13T15:06:15.712Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19041
```

[ESLint Extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) `v2.1.13`

**Comment by bradzacher** [comment]
This is working as expected.
As mentioned in the docs:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#selector-options

> The ordering of selectors does not matter. The implementation will automatically sort the selectors to ensure they match from most-specific to least specific. It will keep checking selectors in that order until it finds one that matches the name.

The `filter` makes it the _highest_ priority config for `property`
```json
{
  "selector": "property",
  "format": ["strictCamelCase"],
  "leadingUnderscore": "allow",
  "filter": {
    "regex": "[- ]",
    "match": false
  }
},
```

Highest priority means that this config is checked first, before this config, which has no `filter`
```json
{
  "selector": "property",
  "modifiers": ["readonly"],
  "format": ["strictCamelCase", "UPPER_CASE"]
},
```

Why is this the case? Because filters are a very specific thing we want to ensure that they are checked before anything else so that they are guaranteed to check and match.
If we didn't give them a high priority, then you would potentially have a filter that never even gets checked because some other (more general) selector matches before it.

**Comment by infacto** [comment]
Thanks for reply. So it's not possible to have [quotes-keys](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#ignore-properties-that-require-quotes) or having additional rules for the same selector? 🤔 It may work as expected. But for me, it's a bug. Because I expect the filter only for this specific rule (selector + modifier). ... How would you achieve this? Impossible?

Maybe a new feature request will help in this case. Like `leadingUnderscore` but for quote-keys. An option to ignore keys which are quoted like `{ takeCase: true, 'not-care': false }`. 

I have to solve the issue: 
- Regular property, format: strictCamelCase, ignore quoted properties.
- Readonly property, format: strictCamelCase and UPPER_CASE.

My current alternative:
- Remove the filter for quoted keys and use ignore comment for this cases.
- Allow UPPER_CASE for all properties.
Both workarounds make the naming convention less useful and result in poorer code. Especially in a team. It's pity.

The filter should scoped only for the particular selector and modifiers. It can have a higher prio depending on the match attribute. I just ask sincerely. Maybe there is a solution today. It feels weird that this is the correct behavior. So that the filter globally makes all other selectors invalid.

**Comment by infacto** [comment]
@bradzacher  Sorry for bother you. ... I could file a new issue about quoted keys. But will be closed for sure. Because I have to use the filter feature. And then I come back to this issue again. Vicious circle. Do you understand my issue? You wrote "This is working as expected." But it means "it works like it works", but not as it should be work. Please tell me, how to achieve that (described above). It's something that typescript-eslint supports. But actually not. This is finally a bug. The filter should only affected to the defined object (selector and modifier) in naming-conventions. Either there is a solution for this issue or now. Ignoring does not solve it. 

Maybe this is actually an quoted keys issue. We should find a way to handle this better. Instead of using the filter. Would it possible that this linter handles the quotes around key names instead of filtering the key names itself? In best case we have an option to ignore quotes keys or handle it in another way. Maybe with a modifier. To be allowed of using quotes also on key names without characters like `- _` (separators). Any guess or hint will help me. 🙂
